File: trunk/cplusplus/src/trees/DecisionTree.cpp

#include "trees/DecisionTree.h"
#include "trees/DecisionNode.h"
#include "algorithms/mode.h"
#include "definitions.h"
#include "utility/print_utils.h"
#include "algorithms/index_related.h"
#include "utility/eigen_helper.h"
#include "utility/errors.h"
#include <algorithm>
#include <functional>
#include <memory>
using namespace Eigen;
using namespace ml::algorithms;
using namespace ml::utility;        

namespace ml {
namespace trees {


void DecisionTree::do_training(const MatrixXd &data, const VectorXi &classes,
              VariableTypes variable_types) {
  columns_to_use_.resize(data.cols());
  std::fill(columns_to_use_.begin(), columns_to_use_.end(), true);
  variable_types_ = variable_types;
  DecisionNodePtr p(new DecisionNode());
  // Casting is neccessary because set_root expects a TreeNodePtr
  set_root(std::dynamic_pointer_cast<TreeNode>(p));
  get_tree(p, data, classes);
}


void DecisionTree::get_tree(DecisionNodePtr node,
                            const MatrixXd &data,
                            const VectorXi &classes) {
  
  IntsSet unique_classes(classes.data(), classes.data()+classes.size());
  if(unique_classes.size() == 1 || number_of_columns_in_use() == 1) {
    IntPair mode_pair = mode(classes.data(), classes.size());
    node->set_class(mode_pair.first);
  } else {
    // (gain, threshold, column for split)
    std::tuple<double, unsigned int, double> gain = get_best_gain(data, classes);
    double g  = std::get<0>(gain);
    double threshold = std::get<2>(gain);
    unsigned int column = std::get<1>(gain);
    node->set_column_for_next_split(column);
    columns_to_use_[column] = false;
    if(variable_types_[column] == CATEGORICAL) {
      IntsSet vals;
      for(unsigned int i = 0; i < data.rows(); i++) {
        vals.insert(data.cast<int>()(i,column));
      }
      for(auto value: vals) {
        DecisionNodePtr child(new DecisionNode());
        child->set_feature(column);
        child->set_feature_value(value);
        node->add_child(child);
        Ints new_rows_to_use = rows_where_true<int>(data, column,
                std::bind2nd(std::equal_to<int>(), value));
//        print(new_rows_to_use.begin(), new_rows_to_use.end(), std::cout);
        // TODO: Find a way of selecting rows without copying them 
        // (some sort of view?)
        MatrixXd new_data = select_rows<double, Ints>(data, new_rows_to_use);
        VectorXi new_classes = select_rows<int, Ints>(classes, new_rows_to_use);
        get_tree(child, new_data, new_classes);
      }
    // continuous feature  
    } else {
      std::vector<ThresholdType> types = {HIGHER, LOWER};
      for(auto type: types) {
        DecisionNodePtr child(new DecisionNode());
        child->set_feature(column);
        child->set_feature_value(threshold);
        child->set_threshold_type(type);
        node->add_child(child);
        Ints new_rows_to_use;
        if(type == LOWER) {
          new_rows_to_use = rows_where_true<double>(data, column,
                std::bind2nd(std::less_equal<double>(), threshold));
        } else if(type == HIGHER) {
          new_rows_to_use = rows_where_true<double>(data, column,
                std::bind2nd(std::greater<double>(), threshold));
        }
        MatrixXd new_data = select_rows<double, Ints>(data, new_rows_to_use);
        VectorXi new_classes = select_rows<int, Ints>(classes, new_rows_to_use);
        get_tree(child, new_data, new_classes);
      }
    }
  }
}  

unsigned int DecisionTree::number_of_columns_in_use() const {
  return std::count(columns_to_use_.begin(), columns_to_use_.end(), true);
}


VectorXi DecisionTree::get_prediction(const MatrixXd &data) const {
  VectorXi predictions = VectorXi::Zero(data.rows());
  for(std::size_t i = 0; i < data.rows(); i++) {
    DecisionNodePtr p = std::dynamic_pointer_cast<DecisionNode>(root_);
    predictions(i) =  get_prediction(p.get(), data.row(i));
  }
  return predictions;
}

int DecisionTree::get_prediction_datapoint(const VectorXd &data) const {
  DecisionNodePtr p = std::dynamic_pointer_cast<DecisionNode>(root_);
  int pred = get_prediction(p.get(), data);
  return pred;
}


int DecisionTree::get_prediction(DecisionNode *node, 
                            const VectorXd &data_point) const {
  int prediction = 0;
  if(node->is_leaf()) {
    return node->get_class();
  } else {
    unsigned int col = node->get_column_for_next_split();
    NodePtr child = node->get_first_child();
    while(child != 0) {
      // Recover a pointer to DecisionNode. It can be done because I know that
      // children are indeed DecisionNodes and the class TreeNode is polymorphic.
      // No need to delete (there is no memory allocation involved
      DecisionNodePtr ptr = std::dynamic_pointer_cast<DecisionNode>(child);
      if(ptr->matches_value(data_point(col), variable_types_[col])) {
        return get_prediction(ptr.get(), data_point);
      }   
      child = child->get_next_sibling();
    }
    throw ValueError("Value for the node not found in the tree");
  }
}

std::tuple<double, unsigned int, double>
  DecisionTree::get_best_gain(const MatrixXd &data, 
                              const VectorXi &classes) {
  // If best_gain == 0 and all the gains for the columns are 0, the 
  // best feature is always set to 0. Avoid it setting best-gain to -1
  double best_gain = -1; 
  double threshold = 0;
  double best_feature = 0;
  for(int i=0; i < columns_to_use_.size(); ++i) {
    if(columns_to_use_[i]) {
      GainPair g = information_gain(data.col(i), classes, variable_types_[i],
                                                         information_measure_);
      if(g.first > best_gain) { 
        best_gain = g.first;
        threshold = g.second;
        best_feature = i;
      }
    }
  }
  auto result = std::make_tuple(best_gain, best_feature, threshold);
  return result;
}
} // trees
} // ml

File: trunk/matlab/trees/DecisionTree.m
classdef DecisionTree < Tree & handle
    properties
        gain_measure;
        % columns that are used for each step during building the tree
        columns_in_use;
        % vector indicating for each feature if it is categorical (true)
        % or not (false).
        are_categorical;
    end
    
    
    
    methods
        function self = DecisionTree()
            root = DecisionNode;
            self = self@Tree(root);
            self.gain_measure = InfoMeasure.gini; % it can be also 'entropy'
        end
        
        function self = set.gain_measure(self, m)
            self.gain_measure = m;
        end
            
        %{
           Train the tree.
            - data is expected to be a real matrix of datapoints (rows)
                and features (cols)
            - Y the expected class for each datapoint
            - are_categorical is a vector with 0 for a continuous 
              variable and 1 for a categorical one
        %} 
        function self = train(self, X, Y, are_categorical)
            self.columns_in_use = ones(1,size(X,2));
            self.are_categorical = are_categorical;
            root = DecisionNode;
            self.root = self.get_tree(root, X, Y);            
        end
        
        
        function tree = get_tree(self, node, X, Y)
            if (number_of_categories(Y) == 1 ||  ...
                                    self.number_of_columns_in_use() == 1)
                node.class = get_class(Y);
                tree = node;
            else
                [gain, column, threshold] = self.best_gain(X, Y);
                node.column_for_next_split = column;
                self.columns_in_use(column) = false;
                is_categorical = self.are_categorical(column);
                if is_categorical
                    values = unique(X(:, column));
                    n_values = size(values,1);
                    for i = 1:n_values
                        [Z, W] = get_matrices_for_value(X, Y,...
                                                        column, values(i));
                        child = DecisionNode;
                        child.set_feature(column, values(i));
                        child_tree = self.get_tree(child, Z, W);
                        
                        node.add_child(child_tree);
                    end
                else
                    h = [InfoGainThreshold.higher InfoGainThreshold.lower];
                    for i = 1:2
                        [Z, W] = get_matrices_for_threshold( ...
                                        X, Y, column, threshold, h(i));
                        child = DecisionNode;
                        child.set_feature(column, threshold);
                        child.threshold_type = h(i);
                        child_tree = self.get_tree(child, Z, W);
                        node.add_child(child_tree);
                    end
                end
                tree = node;         
            end
        end
        
               
        function [bestGain, best_feature, threshold] = best_gain(self, X, Y)
            cols = size(self.columns_in_use,2);
            bestGain = 0; % best gain
            threshold = 0;
            best_feature = 0;
            for i = 1:cols
                if(self.columns_in_use(i))
                    [g, thr] = info_gain(X(:, i), Y, ...
                            self.are_categorical(i), self.gain_measure);
                    if(g > bestGain)
                        bestGain = g; 
                        best_feature = i;
                        threshold = thr;
                    end
                end
            end
        end

        
        function n = number_of_columns_in_use(self)
            n = sum(self.columns_in_use);
        end

        
        
        function show(self)
            show@Tree(self);
        end
        
        
        function predictions = predict(self, data)
           %{
                Predict the class in Y for each of the datapoints in X
                Each row is a datapoint
            %}
           n = size(data,1);
           predictions = zeros(n, 1);
           for i = 1:n
               predictions(i, 1) = self.get_prediction(self.root, data(i, :));
           end
            
        end
            
        function prediction = get_prediction(self, node, x)
            %{
                returns the predicted class for this node.
                x is a row vector datapoint 
            %}
            if(node.isleaf())
                prediction = node.class;
                %sprintf('I have a prediction for x => %s\n',char(prediction))
                return 
            else
                col = node.column_for_next_split;
                value_xcol = x(col);
                children = node.get_children();
                n_children = node.get_number_of_children();
                for i = 1:n_children
                    if(children(i).matches_value(x(col), ...
                                        self.are_categorical(col)))
                        prediction = self.get_prediction(children(i), x);
                        return
                    end
                end
                exception = MException('DecisionTree:get_prediction', ...
                   'Value for the node not found in the tree');
                throw(exception)
            end
        end
    end
end
           
    
    function n_cats = number_of_categories(Y)
        n_cats = size(unique(Y),1);
    end
       
    % class is chosen as the value that appears more times
    function class = get_class(Y)
        counts = levelcounts(Y);
        levels = getlevels(Y);
        [max_val, index] = max(counts);
        class = levels(index);
    end
    
    
    % get the datapoints in matrices X and Y where the column "column"
    % in X has the value "value"
    function [Z, W] = get_matrices_for_value(X, Y, column, value)
        indices = (X(:, column) == value);
        Z = X(indices, :);
        W = Y(indices, :);
    end
        
    % Same as previous function but for values lower or higher
    function [Z, W] = get_matrices_for_threshold(X, Y, column, ...
                                            threshold, lower_or_higher)                                        
        if(lower_or_higher == InfoGainThreshold.lower)
            indices = (X(:, column) <= threshold);
        else
            indices = (X(:, column) > threshold);
        end
        Z = X(indices, :);
        W = Y(indices, :);
    end
        

 
